### PR TITLE
Failing test for GH8443

### PR DESCRIPTION
**Pull request by piowin**
A test suit for #8443.

Closes #8443

### CODE REVIEW

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+use function count;
+
+final class GH8443Test extends OrmFunctionalTestCase
+{
+    /**
+     * @group GH-8443
+     */
+    public function testJoinRootEntityWithForcePartialLoad(): void
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php:33]: Running that test by itself fails with `SQLSTATE[HY000]: General error: 1 no such table: company_persons`.
Tests should be independent from each other

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /**
+     * @group GH-8443
+     */
+    public function testJoinRootEntityWithOnlyOneEntityInHierarchy(): void
+    {
+        try {
+            $this->_schemaTool->createSchema([$this->_em->getClassMetadata(GH8443Foo::class)]);
+        } catch (Exception $e) {
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php]: There should be an import for this class

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        $manager->setTitle('manager');
+
+        $manager->setSpouse($person);
+
+        $this->_em->persist($person);
+        $this->_em->persist($manager);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $result = $this->_em->createQuery("select m from Doctrine\Tests\Models\Company\CompanyManager m JOIN m.spouse s WITH s.name = 'John'")->setHint(Query::HINT_FORCE_PARTIAL_LOAD, true)->getResult();
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php]: Why not use `getSingleResult`?

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        $this->assertEquals(count($result), 1);
+
+        $manager = $result[0];
+        $this->_em->refresh($manager);
+
+        $this->assertEquals($manager->getSpouse()->getName(), 'John');
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH8443Test.php]:
```suggestion
        $this->assertEquals('John', $manager->getSpouse()->getName());
```

**Comment by greg0ire** [comment]
> The same problem appears if you add WITH condition to the joined entity with discriminator

As opposed to what? https://github.com/doctrine/orm/pull/6812 is also about `WITH`, so I'm confused…

**Comment by piowin** [comment]
> 
> 
> > The same problem appears if you add WITH condition to the joined entity with discriminator
> 
> As opposed to what? #6812 is also about `WITH`, so I'm confused…

You're right. In both issues the problem was a result of using a with condition. I think the code of SqlWalker::walkJoinAssociationDeclaration could be simplfy a little and the #8443 wouldn't be necessary, but then the code doesn't pass some tests of the SelectSqlGeneratorTest, even though the sql syntax is correct.

**Comment by greg0ire** [comment]
I'll rephrase my question: what is the difference between both issues? Since the first patch did not fix the second issue, there must be some difference somewhere, right?

**Comment by piowin** [comment]
The sqls are generated in different methods with different logic.

**Comment by greg0ire** [comment]
Ok, why is that?

**Comment by piowin** [comment]
I haven't designed the SqlWalker, but I see that, in the terms of the EBNF of DQL, the logic of the generation of Join is more complex when we join JoinAssociationDeclaration than when we join RangeVariableDeclaration . Some of the logic is similar and maybe it would be possible to put that logic in a separate method.

**Comment by greg0ire** [comment]
> when we join JoinAssociationDeclaration than when we join RangeVariableDeclaration

Ah, so that's the difference! How do each of these look like in DQL?

**Comment by piowin** [comment]
A JoinAssociationDeclaration  is a path to an entity's field (the field can be a collection of entities, a single entity or a simple state field) plus an alias: p.friend AS f (p must be defined before and that's one of the differences with the next one).

A RangeVariableDeclaration is a path to a class plus an alias: \MyNamespace\Person AS p. A RangeVariableDeclaration is the obligatory component that must be placed immediately after the "FROM" clause.

A JoinAssociationDeclaration  can be only used as a joined component while A RangeVariableDeclaration can be used as a joined component and a starting point/base of an IdentificationVariableDeclaration which is a building block of the "FROM" clause.

How does the method responsible for generating joins work?
When the sql for a Join is being generated, if the Join is a JoinAssociationDeclaration, it simply delegates to the method that is responsible for generating JoinAssociationDeclaration, that's all. But if the Join is a RangeVariableDeclaration, it generates optional conditions for the RangeVariableDeclaration, like a filter condition, a "with" condition or a discriminator condition. The generation of each conidtion has its unique logic. Also the sql for the RangeVariableDeclaration is being generated by the method that is responsible for that (the method is invoked also when we genarate IdentificationVariableDeclaration, see above). The conditions are added to the sql and the result is returned.

The logic of the generation of the optional brackets (the subject of our discussion) is put in the method that is reponsible for generating JoinAssociationDeclaration, if the Join is a JoinAssociationDeclaration, and is put in the method that is responsible for generating joins, if the Join is a RangeVariableDeclaration.

**Review by greg0ire — APPROVED** [review]
Very clear! Thank you!

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @piowin!

**Pull request merged**